File: src/90_subsets_ii/subsets_ii.py
from typing import List


class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        def backtrack(start, path):
            res.append(path[:])
            for i in range(start, len(nums)):
                if i > start and nums[i] == nums[i - 1]:
                    continue
                path.append(nums[i])
                backtrack(i + 1, path)
                path.pop()

        backtrack(0, [])
        return res


# Test scaffolding
if __name__ == "__main__":
    sol = Solution()
    print(
        sorted([sorted(x) for x in sol.subsetsWithDup([1, 2, 2])]),
        "|",
        sorted([sorted(x) for x in [[], [1], [1, 2], [1, 2, 2], [2], [2, 2]]]),
    )
    print(
        sorted([sorted(x) for x in sol.subsetsWithDup([0])]),
        "|",
        sorted([sorted(x) for x in [[], [0]]]),
    )
    print(sorted([sorted(x) for x in sol.subsetsWithDup([4, 4, 4, 1, 4])]))

File: src/90_subsets_ii/subsets_ii.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        sort(nums.begin(), nums.end());
        backtrack(nums, 0, path, res);
        return res;
    }

   private:
    void backtrack(vector<int>& nums, int start, vector<int>& path, vector<vector<int>>& res) {
        res.push_back(path);
        for (int i = start; i < nums.size(); ++i) {
            if (i > start && nums[i] == nums[i - 1]) continue;
            path.push_back(nums[i]);
            backtrack(nums, i + 1, path, res);
            path.pop_back();
        }
    }
};

// Test scaffolding
int main() {
    Solution sol;
    vector<int> nums1 = {1, 2, 2};
    vector<vector<int>> res1 = sol.subsetsWithDup(nums1);
    for (auto& v : res1) {
        cout << "[";
        for (int n : v) cout << n << ",";
        cout << "] ";
    }
    cout << endl;
    vector<int> nums2 = {0};
    vector<vector<int>> res2 = sol.subsetsWithDup(nums2);
    for (auto& v : res2) {
        cout << "[";
        for (int n : v) cout << n << ",";
        cout << "] ";
    }
    cout << endl;
    return 0;
}
